torch/fx/OVERVIEW.md

```diff
@@ -0,0 +1,121 @@
+
+```python
+import torch
+
+class MyModule(torch.nn.Module):
+  def __init__(self):
+    super().__init__()
+    self.param = torch.nn.Parameter(
+        torch.rand(3, 4))
+    self.linear = torch.nn.Linear(4, 5)
+
+  def forward(self, x):
+    return self.linear(x + self.param).clamp(min=0.0, max=1.0)
+
+from torch.fx import symbolic_trace
+module = MyModule()
+symbolic_traced : torch.fx.GraphModule = symbolic_trace(module)
+
+input = torch.rand(3, 4)
+torch.testing.assert_allclose(symbolic_traced(input), module(input))
+```
+
+Here, we set up a simple Module that exercises different language features: fetching a parameter, applying an arithmetic operator, applying a submodule (linear), and applying a Tensor method. `symbolic_trace` returns an instance of GraphModule, which is in itself a subclass of `nn.Module`. We can see that the `symbolic_traced` instance runs and returns the same result as the original module instance module.
+
+## Tracer ##
+
+`Tracer` is the class that implements the symbolic tracing functionality of `torch.fx.symbolic_trace`. A call to `symbolic_trace(m)` is equivalent to `Tracer().trace(m)`. Tracer can be subclassed to override various behaviors of the tracing process. The different behaviors that can be overridden are described in the docstrings of the methods on the class.
+
+In the default implementation of `Tracer().trace`, the tracer first creates Proxy objects for all arguments in the `forward` function. (This happens in the call to `create_args_for_root`.) Next, the `forward` function is called with the new Proxy arguments. As the Proxies flow through the program, they record all the operations (`torch` function calls, method calls, and operators) that they touch into the growing FX Graph as Nodes.
+
+## Proxy ##
+
+Proxy objects are Node wrappers used by the Tracer to record operations seen during symbolic tracing. If you're doing graph transforms, you can wrap your own Proxy method around a raw Node so that you can use the overloaded operators to add additional things to a Graph.
+
+The mechanism through which Proxy objects record computation is [`__torch_function__`](https://pytorch.org/docs/stable/notes/extending.html). If any custom Python type defines a method named `__torch_function__`, PyTorch will invoke that `__torch_function__` implementation when an instance of that custom type is passed to a function in the `torch` namespace. In the Proxy implementation of `__torch_function__`, we simply create a new Proxy object with the passed-in operator. Creating a new Proxy means that a new Node--the Node wrapped by the Proxy--will be added to the Graph as well. When this new Proxy object is used later in the code, it “transforms” its user into another Proxy object and the process is repeated.
```

Comment:
For the `__torch_function__` link I recommend using the anchor link (https://pytorch.org/docs/stable/notes/extending.html#extending-torch) to go directly to the `__torch_function__` section

Reply:
In the prose, I think the causality is sort-of reversed. I'd describe the process as "1) operations on Proxy are dispatched to the `__torch_function__` handler 2) The `__torch_function__` handler records the operation in the `Graph` as a `Node` 3) The `Node` that was recorded in the `Graph` is wrapped in a `Proxy`, facilitating further application of ops on that value"